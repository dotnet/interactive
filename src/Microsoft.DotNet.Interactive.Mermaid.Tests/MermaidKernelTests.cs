// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Assent;
using FluentAssertions;
using FluentAssertions.Execution;
using HtmlAgilityPack;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive.Tests.Utility;
using Xunit;

namespace Microsoft.DotNet.Interactive.Mermaid.Tests;

public class MermaidKernelTests 
{
    private readonly Configuration _configuration;
    

    public MermaidKernelTests()
    {

        _configuration = new Configuration()
            .UsingExtension("txt")
            .SetInteractive(Debugger.IsAttached);
    }
    
    [Fact]
    public void registers_html_formatter_for_MermaidMarkdown()
    {
        var markdown = @"graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]";

        var formatted = new MermaidMarkdown(markdown).ToDisplayString(HtmlFormatter.MimeType);
        var doc = new HtmlDocument();
        doc.LoadHtml(formatted.FixedGuid().FixedCacheBuster());
        var scriptNode = doc.DocumentNode.SelectSingleNode("//div/script");
        var renderTarget = doc.DocumentNode.SelectSingleNode("//div[@id='00000000000000000000000000000000']");
        using var _ = new AssertionScope();

        scriptNode.Should().NotBeNull();
        scriptNode.InnerText.Should()
            .Contain(markdown);
        scriptNode.InnerText.Should()
            .Contain("(['mermaidUri'], (mermaid) => {");

        renderTarget.Should().NotBeNull();
    }

    [Fact]
    public async Task mermaid_kernel_handles_SubmitCode()
    {
        using var kernel = new CompositeKernel
        {
            new MermaidKernel()
        };

        var result = await kernel.SubmitCodeAsync(@"
#!mermaid
    graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]
");

        var events = result.KernelEvents.ToSubscribedList();

        var returnValue = events
            .OfType<DisplayedValueProduced>()
            .Single()
            .Value as MermaidMarkdown;

        returnValue.ToString().Should().Be(@"graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]
");
        
    }

    [Fact]
    public async Task can_specify_background_color()
    {
        using var kernel = new CompositeKernel
        {
            new MermaidKernel()
        };

        KernelCommandResult result = await kernel.SubmitCodeAsync(@"
#!mermaid --display-background-color red
    graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]
");

        var events = result.KernelEvents.ToSubscribedList();

        var formattedData = events
            .OfType<DisplayedValueProduced>()
            .Single()
            .FormattedValues
            .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
            .Value;

        var htmlDoc = new HtmlDocument();
        htmlDoc.LoadHtml(formattedData.FixedGuid());

        var style = htmlDoc.DocumentNode
                .SelectSingleNode("//div[contains(concat(' ',normalize-space(@class),' '),' mermaidMarkdownContainer ')]")
                .Attributes["style"].Value;

        style.Should().Be("background-color:red");
    }

    [Fact]
    public async Task can_specify_display_dimensions()
    {
        using var kernel = new CompositeKernel
        {
            new MermaidKernel()
        };

        KernelCommandResult result = await kernel.SubmitCodeAsync(@"
#!mermaid --display-width 200px --display-height 250px
    graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]
");

        var events = result.KernelEvents.ToSubscribedList();

        var formattedData = events
            .OfType<DisplayedValueProduced>()
            .Single()
            .FormattedValues
            .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
            .Value;

        var htmlDoc = new HtmlDocument();
        htmlDoc.LoadHtml(formattedData.FixedGuid());

        var style = htmlDoc.DocumentNode
            .SelectSingleNode("//div[@id='00000000000000000000000000000000']")
            .Attributes["style"].Value;

        style.Should().Be(" width:200px;  height:250px; ");

    }
}