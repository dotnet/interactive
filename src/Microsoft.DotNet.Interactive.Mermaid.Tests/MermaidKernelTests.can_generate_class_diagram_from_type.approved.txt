classDiagram

class JsonElement
JsonElement : -CheckValidInstance() Void
JsonElement : +Clone() JsonElement
JsonElement : +EnumerateArray() ArrayEnumerator
JsonElement : +EnumerateObject() ObjectEnumerator
JsonElement : -get_DebuggerDisplay() String
JsonElement : +get_Item(Int32 index) JsonElement
JsonElement : -get_TokenType() JsonTokenType
JsonElement : +get_ValueKind() JsonValueKind
JsonElement : +GetArrayLength() Int32
JsonElement : +GetBoolean() Boolean
JsonElement : +GetByte() Byte
JsonElement : +GetBytesFromBase64() Byte[]
JsonElement : +GetDateTime() DateTime
JsonElement : +GetDateTimeOffset() DateTimeOffset
JsonElement : +GetDecimal() Decimal
JsonElement : +GetDouble() Double
JsonElement : +GetGuid() Guid
JsonElement : +GetInt16() Int16
JsonElement : +GetInt32() Int32
JsonElement : +GetInt64() Int64
JsonElement : +GetProperty(String propertyName) JsonElement
JsonElement : +GetProperty(ReadOnlySpan~Char~ propertyName) JsonElement
JsonElement : +GetProperty(ReadOnlySpan~Byte~ utf8PropertyName) JsonElement
JsonElement : GetPropertyName() String
JsonElement : GetPropertyRawText() String
JsonElement : +GetRawText() String
JsonElement : GetRawValue() ReadOnlyMemory~Byte~
JsonElement : +GetSByte() SByte
JsonElement : +GetSingle() Single
JsonElement : +GetString() String
JsonElement : +GetUInt16() UInt16
JsonElement : +GetUInt32() UInt32
JsonElement : +GetUInt64() UInt64
JsonElement : +ParseValue(Utf8JsonReader& reader) JsonElement$
JsonElement : ParseValue(Stream utf8Json, JsonDocumentOptions options) JsonElement$
JsonElement : ParseValue(ReadOnlySpan~Byte~ utf8Json, JsonDocumentOptions options) JsonElement$
JsonElement : ParseValue(String json, JsonDocumentOptions options) JsonElement$
JsonElement : TextEqualsHelper(ReadOnlySpan~Byte~ utf8Text, Boolean isPropertyName, Boolean shouldUnescape) Boolean
JsonElement : TextEqualsHelper(ReadOnlySpan~Char~ text, Boolean isPropertyName) Boolean
JsonElement : +ToString() String
JsonElement : +TryGetByte(Byte& value) Boolean
JsonElement : +TryGetBytesFromBase64(Byte[]& value) Boolean
JsonElement : +TryGetDateTime(DateTime& value) Boolean
JsonElement : +TryGetDateTimeOffset(DateTimeOffset& value) Boolean
JsonElement : +TryGetDecimal(Decimal& value) Boolean
JsonElement : +TryGetDouble(Double& value) Boolean
JsonElement : +TryGetGuid(Guid& value) Boolean
JsonElement : +TryGetInt16(Int16& value) Boolean
JsonElement : +TryGetInt32(Int32& value) Boolean
JsonElement : +TryGetInt64(Int64& value) Boolean
JsonElement : +TryGetProperty(String propertyName, JsonElement& value) Boolean
JsonElement : +TryGetProperty(ReadOnlySpan~Char~ propertyName, JsonElement& value) Boolean
JsonElement : +TryGetProperty(ReadOnlySpan~Byte~ utf8PropertyName, JsonElement& value) Boolean
JsonElement : +TryGetSByte(SByte& value) Boolean
JsonElement : +TryGetSingle(Single& value) Boolean
JsonElement : +TryGetUInt16(UInt16& value) Boolean
JsonElement : +TryGetUInt32(UInt32& value) Boolean
JsonElement : +TryGetUInt64(UInt64& value) Boolean
JsonElement : +TryParseValue(Utf8JsonReader& reader, Nullable element) Boolean$
JsonElement : +ValueEquals(String text) Boolean
JsonElement : +ValueEquals(ReadOnlySpan~Byte~ utf8Text) Boolean
JsonElement : +ValueEquals(ReadOnlySpan~Char~ text) Boolean
JsonElement : +WriteTo(Utf8JsonWriter writer) Void


